mmdet/models/detectors/soft_teacher.py

```diff
@@ -0,0 +1,386 @@
+                    self.semi_train_cfg.pseudo_label_initial_score_thr]
+
+        reg_uncs_list = self.compute_uncertainty_with_aug(
+            x, batch_data_samples)
+
+        for data_samples, reg_uncs in zip(batch_data_samples, reg_uncs_list):
+            data_samples.gt_instances['reg_uncs'] = reg_uncs
+            data_samples.gt_instances.bboxes = bbox_project(
+                data_samples.gt_instances.bboxes,
+                torch.from_numpy(data_samples.homography_matrix).inverse().to(
+                    self.data_preprocessor.device), data_samples.ori_shape)
+
+        batch_info = {
+            'feat': x,
+            'img_shape': [],
+            'homography_matrix': [],
+            'metainfo': []
+        }
+        for data_samples in batch_data_samples:
+            batch_info['img_shape'].append(data_samples.img_shape)
+            batch_info['homography_matrix'].append(
+                torch.from_numpy(data_samples.homography_matrix).to(
+                    self.data_preprocessor.device))
+            batch_info['metainfo'].append(data_samples.metainfo)
+        return batch_data_samples, batch_info
+
+    def unsup_rpn_loss(self, x: Tuple[Tensor],
+                       batch_data_samples: SampleList) -> dict:
+        """Calculate unsupervised rpn loss from a batch of inputs and pseudo
+        data samples.
+
+        Args:
+            x (tuple[Tensor]): Features from FPN.
+            batch_data_samples (List[:obj:`DetDataSample`]): The batch
+                data samples. It usually includes information such
+                as `gt_instance` or `gt_panoptic_seg` or `gt_sem_seg`,
+                which are `pseudo_instance` or `pseudo_panoptic_seg`
+                or `pseudo_sem_seg` in fact.
+        Returns:
+            dict: A dictionary of rpn loss components
+        """
+
+        rpn_data_samples = copy.deepcopy(batch_data_samples)
+        for data_samples in rpn_data_samples:
+            if data_samples.gt_instances.bboxes.shape[0] > 0:
+                data_samples.gt_instances = data_samples.gt_instances[
+                    data_samples.gt_instances.scores >
+                    self.semi_train_cfg.rpn_pseudo_thr]
+
+        proposal_cfg = self.student.train_cfg.get('rpn_proposal',
+                                                  self.student.test_cfg.rpn)
+        # set cat_id of gt_labels to 0 in RPN
+        for data_sample in rpn_data_samples:
+            data_sample.gt_instances.labels = \
+                torch.zeros_like(data_sample.gt_instances.labels)
+
+        rpn_losses, rpn_results_list = self.student.rpn_head.loss_and_predict(
+            x, rpn_data_samples, proposal_cfg=proposal_cfg)
+        for key in rpn_losses.keys():
+            if 'loss' in key and 'rpn' not in key:
+                rpn_losses[f'rpn_{key}'] = rpn_losses.pop(key)
+        return rpn_losses, rpn_results_list
+
+    def unsup_rcnn_cls_loss(self, x: Tuple[Tensor],
```

Comment:
should we also use loss_rcnn_cls_by_pseudo_labels?

Reply:
rcnn_cls_loss_by_pseudo_instances ?